Support for instrumentation in Phoenix

whatyouhide opened the pull request:
Let me know what you think, I'm pretty happy with how this is already coming out. :D :tada:

josevalim commented:
Great job @whatyouhide! I think we can improve this first iteration by doing more and less work at compile time at the same time.

In particular, the generation of clauses can be done at compile time but all the instrumentation work at runtime. Here is some pseudo code:

```elixir
defmacro instrument(event, runtime \\ [], fun) do
  compile = compile_opts(__CALLER__)
  quote do
    Phoenix.Endpoint.instrument(unquote(event), unquote(compile), unquote(runtime), unquote(fun))
  end
end

# Now for each event in any of the modules, we will generate
def instrument(:render_view, compile, runtime, fun) do
  res0 = Mod0.render_view(:start, compile, runtime)
  res1 = Mod1.render_view(:start, compile, runtime)
  res2 = Mod2.render_view(:start, compile, runtime)

  start = current_time()
  try do
    fun.()
  after
    stop = current_time()
    diff = time_diff(start, stop)
    Mod0.render_view(:stop, diff, res0)
    Mod1.render_view(:stop, diff, res0)
    Mod2.render_view(:stop, diff, res0)
  end
end

# catch all - nobody is interested in those events
def instrument(_event, _compile, _runtime, fun) do
  fun.()
end

if erlang monotonic time do
  defp current_time do
    ...
  end

  defp time_diff do
    ...
  end
else
  ...
end
```

josevalim commented:
Also, let's move the code to `Phoenix.Endpoint.Instrument`. It is preferable to call `Phoenix.Endpoint.Instrument.definstrument(app)` directly from `Phoenix.Endpoint` (it could be during before before_compile hook) than to `use Phoenix.Instrument`. We rely a lot on this pattern on Phoenix (check the endpoint and the router) as it makes it explicit and we don't need to chase before_compile using hooks through different files.

whatyouhide commented:
So the cost of wrapping the code inside a `fn` is near to zero? Because with the way you proposed, when we nobody is interested in an event, then we still call `fun.()`; with my version, if nobody is interested in an event, absolutely nothing happens as far as instrumentation goes.

I have to say this version looks cleaner though, so it may be worth it :) (and I actually remember you saying on Skype that just calling `fun.()` as almost no cost, not sure though).

josevalim commented:
It is very low cost, we don't need to worry about it. So yes, let's just
ask them to pass functions.

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

whatyouhide commented:
@josevalim ready for review, I implemented all the changes you mentioned. I had to rely on some public (but with `@doc false`) functions in `Phoenix.Endpoint.Instrument` in order to clean up `__using__` a little bit.

Also, do we need to enforce that `runtime` is a map? I think it could be just any term, e.g., with `:render_view` it could just be `"index.html"`. Right now, this is what I implemented but changing it back to  an enforced map would take just a couple of minutes.

Let me know what you think!

chrismccord commented:
This looks great @whatyouhide. Would it make any sense to map the results as a list, then map again to report the results instead of building the intermediate vars? Not sure if there are perf implications, but it might make the code a tad simpler. Either way the existing code very very nice

josevalim commented:
> @josevalim ready for review, I implemented all the changes you mentioned. I had to rely on some public (but with @doc false) functions in Phoenix.Endpoint.Instrument in order to clean up __using__ a little bit.

Feel free to have as many functions in `Phoenix.Endpoint.Instrument`, you can even `@doc` them as long as `@moduledoc` false. The whole purpose of that module to exist is to define named functions. :)

josevalim commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+defmodule Phoenix.Endpoint.Instrument do
```
`@moduledoc false`

whatyouhide commented:
> Would it make any sense to map the results as a list, then map again to report the results instead of building the intermediate vars? Not sure if there are perf implications, but it might make the code a tad simpler.

@chrismccord I think we settled on not looping over lists but expanding all the necessary stuff at compile time. Yes mapping over the list of calls to get results and looping again on `:stop` may be cleaner, but I think this way is faster?

@josevalim right, I'll make the module "private" and also call `definstrument` on the endpoint.

lexmag commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+      # TODO: maybe it's better to make Phoenix endpoints export __otp_app__()
+      # or something similar?
+      config        = Application.get_env(var!(otp_app), __MODULE__, [])
+      instrumenters = config[:instrumentation] || []
```
Maybe better to use `Keyword.get/3` to stop it from silently working with `instrumentation: nil`?

josevalim replied:
Let's keep the current pattern. Most of the times we don't care about the distinction between those and this code snippet is more concise.

lexmag commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+    inst
+    |> Atom.to_string()
+    |> String.replace(".", "_")
+    |> (fn s -> "res_" <> s end).()
```
:see_no_evil:

josevalim commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+      # The `otp_app` variable is set by Phoenix.Endpoint.
+      # TODO: maybe it's better to make Phoenix endpoints export __otp_app__()
+      # or something similar?
+      config        = Application.get_env(var!(otp_app), __MODULE__, [])
```
You can move all this code inside `events_to_instrumenters`. Maybe call it `app_instrumenters`.

josevalim commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+
+      """
+      defmacro instrument(event, runtime \\ nil, fun) do
+        compile = Macro.escape(__CALLER__)
```
We don't need the whole __CALLER__. It may in fact be too much. Literally this: https://github.com/elixir-lang/elixir/blob/master/lib/logger/lib/logger.ex#L533-L538

(we can even copy the compile time app thing).

josevalim commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+      end
+
+      # Catch-all clause
+      def instrument(_event, _compile, _runtime, fun) do
```
Let's add guards here too!

josevalim commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+  end
+
+  defp instrumenters_and_events(instrumenters) do
+    for inst <- instrumenters, {event, _arity} <- inst.__info__(:functions) do
```
Let's just copy the ones with `arity == 3`

josevalim commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+  # use for that instrumenter module (the result variable is just the result
+  # returned by the start callback of an event).
+  defp build_result_variable(inst) do
+    inst
```
Variables could be simpler. Just receive an integer based on the number of instrumenters and name them res_0, res_1, res_2 and so on. You can use `Enum.with_index` to tie the instrumenters to an index.

josevalim commented on lib/phoenix/endpoint/instrumenter.ex:
```diff
@@ -0,0 +1,138 @@
+  def events_to_instrumenters(instrumenters) do
+    instrumenters                                # [Inst1, Inst2, ...]
+    |> instrumenters_and_events()                # [{Inst1, ev1}, {Inst1, ev2}, {Inst2, ev1}, ...]
+    |> Enum.group_by(fn {_inst, evt} -> evt end) # %{ev1 => [Inst1, Inst2], ev2 => ...}
```
I think the comment is wrong. it will become `%{ev1 => [{Inst1, ev}, ...]}`

whatyouhide replied:
Whoops, yes I mixed that line and the next line up :)

josevalim replied:
I just knew because I was bitten by something similar a couple day ago :P

whatyouhide commented:
@josevalim @chrismccord @lexmag this is ready for review.

I addressed all the comments @josevalim mentioned (stripped the `__CALLER__`, only took events with arity of `3`, used `Instrument.definstrument(var!(otp_app))` in the endpoint, called result variables `res0` ).

I think the code that resulted here is pretty clean, but we may be able to clean it up even further with something like:

```elixir
defmodule Phoenix.Endpoint.Instrument do
  defmacro definstrument(app) do
    quote bind_quoted: [app: app] do
      Phoenix.Endpoint.Instrument.def_instrument_macro()
      Phoenix.Endpoint.Instrument.def_instrument_function(app, __MODULE__)
    end
  end
end
```

but while cleaning up `definstrument/1`, I think this makes the rest slightly harder to follow. Your choice anyways, it shouldn't take long to make these changes :)

Gazler commented on lib/phoenix/endpoint/instrument.ex:
```diff
@@ -0,0 +1,199 @@
+  end
+
+  @doc """
+  Reads a list of the instrumnters from the config of `otp_app` and finds all
```
instrumenters

Gazler commented on lib/phoenix/endpoint/instrument.ex:
```diff
@@ -0,0 +1,199 @@
+  end
+
+  @doc """
+  Strips a `Macro.Env` struct, leaving only intereseting compile-time metadata.
```
interesting not intereseting

Gazler commented:
@whatyouhide I just found a couple of typos in the documentation. Looks great apart from that.

josevalim commented on lib/phoenix/endpoint.ex:
```diff
@@ -226,6 +226,9 @@ defmodule Phoenix.Endpoint do
       unquote(server())
+
+      require Phoenix.Endpoint.Instrument
+      Phoenix.Endpoint.Instrument.definstrument(var!(otp_app))
```
I was thinking more about something like this (inside `__before_compile__`):

```elixir
otp_app = Module.get_attribute(env.module, :otp_app)
instrumentation = Phoenix.Endpoint.Instrument.definstrument(otp_app, env.module)

quote do
  unquote(instrumentation)
end
```

There is no reason to use a macro in this particular case. :)

whatyouhide commented:
@josevalim aaah, ok I see, makes a lot of sense :) Pushed 77a6873 with the new code!
 to

josevalim commented on lib/phoenix/endpoint/instrument.ex:
```diff
@@ -0,0 +1,187 @@
+      end
+
+      # Let's define the current_time/0 and time_diff/2 functions based on the
+      # existence of :erlang.monotonic_time/1.
```
Let's leave a todo: `# TODO: Remove this once Phoenix supports only Elixir 1.2`.

josevalim commented on lib/phoenix/endpoint/instrument.ex:
```diff
@@ -0,0 +1,187 @@
+    app_instrumenters = app_instrumenters(otp_app, endpoint)
+
+    quote bind_quoted: [app_instrumenters: app_instrumenters] do
+      @doc """
```
Those docs are excellent! We should also include a snippet about instrumentation covering all aspects, from configuring up to defining your own module up to instrument in a new section in `Phoenix.Endpoint`.

josevalim commented:
It looks perfect, we just need a bit more of docs and we are ready to go! :D

whatyouhide commented:
@josevalim ok, I added docs in `Phoenix.Endpoint`. Let me know what you think!

josevalim commented:
Ship it! :D Now we need to start making use of this!

whatyouhide commented:
![pz96rxa](https://cloud.githubusercontent.com/assets/3890250/11452142/78bfd952-95dd-11e5-9f4d-a8527c808347.png)

(@josevalim I can't ship it :))

josevalim commented:
@whatyouhide I was going to merge then I checked the build and I think you have a heisentest: https://travis-ci.org/phoenixframework/phoenix/builds/93650805#L377

@chrismccord you need to get on top of those other heisentests :)

whatyouhide commented:
@josevalim ok, possibly fixed the heisentest :)

chrismccord commented:
I thought we fixed our Heisentests with @gazler's patch?


> On Nov 28, 2015, at 9:38 AM, José Valim <notifications@github.com> wrote:
> 
> @whatyouhide I was going to merge then I checked the build and I think you have a heisentest: https://travis-ci.org/phoenixframework/phoenix/builds/93650805#L377
> 
> @chrismccord you need to get on top of those other heisentests :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

josevalim commented on lib/phoenix/endpoint.ex:
```diff
@@ -77,6 +77,10 @@ defmodule Phoenix.Endpoint do
 
       The default format is used when none is set in the connection.
 
+    * `:instrumentation` - a list of instrumenters modules whose callbacks will
```
Shouldn't we call it `instrumenters`? Given the description, that seems to be what makes most sense.

whatyouhide replied:
@josevalim makes sense, pushed a fix for that.

josevalim commented:
@chrismccord see: https://travis-ci.org/phoenixframework/phoenix/builds/93650805#L358
